extensions/amp-next-page/0.1/next-page-service.js

```diff
@@ -269,53 +270,56 @@ export class NextPageService {
       }
 
       this.nextArticle_++;
-      this.xhr_.fetch(next.ampUrl, {ampCors: false})
-          .then(response => {
-            // Update AMP URL in case we were redirected.
-            documentRef.ampUrl = response.url;
-            const url = this.urlService_.parse(response.url);
-            userAssert(url.origin === this.origin_,
-                'ampUrl resolved to a different origin from the origin of the '
-                + 'current document');
-            return response.text();
-          })
-          .then(html => {
-            const doc =
-                this.win_.document.implementation.createHTMLDocument('');
-            doc.open();
-            doc.write(html);
-            doc.close();
-            return doc;
-          })
-          .then(doc => new Promise((resolve, reject) => {
-            if (documentRef.cancelled) {
-              // User has reached the end of the document already, don't render.
-              resolve();
-              return;
-            }
+      const fetchPromise = this.xhr_.fetch(ampUrl, {ampCors: false}).then(r => {
```

Comment:
```suggestion
    this.fetchNextArticle_(next, documentRef)
        .catch(e => user().error(TAG, 'Failed to fetch: %s', ampUrl, e))
        .then(html => {
          if (!html || documentRef.cancelled) {
            // User has reached the end of the document already, don't render.
            return;
          }

          if (documentRef.recUnit.isObserving) {
            this.positionObserver_.unobserve(articleLinks);
            documentRef.recUnit.isObserving = true;
          }

          return this.appendNextArticleHtml_(container, html);
        })
        // The new page may be short and the next may already need fetching.
        .then(() => this.scrollHandler_());
  }

  
  fetchNextArticle(next, documentRef) {
    return this.xhr_.fetch(next.ampUrl, {ampCors: false})
      .then(response => {
        // Update AMP URL in case we were redirected.
        documentRef.ampUrl = response.url;
        const url = this.urlService_.parse(response.url);
        userAssert(url.origin === this.origin_,
            'ampUrl resolved to a different origin from the origin of the '
            + 'current document');
        return response.text();
      });
  }

  appendNextArticleHtml(container, html) {
    const doc = this.win_.document.implementation.createHTMLDocument('');
    doc.open();
    doc.write(html);
    doc.close();

    const shadowRoot = this.win_.document.createElement('div');
    container.appendChild(shadowRoot);

    return this.resources_.mutateElement(container, () => {
      try {
        const amp = this.attachShadowDoc_(shadowRoot, doc);
        documentRef.amp = amp;

        toggle(dev().assertElement(documentRef.recUnit.el), false);
        this.documentQueued_ = false;
      } catch (e) {
        dev().error(TAG,
          'failed to attach shadow document for %s', next.ampUrl, e);
      }
    });
  }
```

Reply:
`appendNextArticleHtml` needs to pass in `documentRef` and `next.ampUrl` too. 

Personally I don't find this more readable e.g. both "fetch" and "append" mutate `documentRef`, so readers need to jump between methods to understand what's happening. That can be fixed with more reordering but it's more than I'd like for the scope of this PR.